File: archive/2019/solutions/MilkyWay.java
import java.io.*;
import java.util.*;

public class MilkyWay {

	@SuppressWarnings("unchecked")
	public static void main(String[] args) throws IOException {
		
		Scanner sc = new Scanner(System.in);
		
		int n = sc.nextInt(), m = sc.nextInt(), c = sc.nextInt();
		int a = sc.nextInt() - 1, b = sc.nextInt() - 1;
		int[] p = new int[n];
		for (int i=0; i<n; i++)
			p[i] = sc.nextInt();
		
		List<Edge>[] adj = new List[n];
		for (int i=0; i<n; i++)
			adj[i] = new LinkedList<Edge>();
		for (int i=0; i<m; i++) {
			int x = sc.nextInt() - 1, y = sc.nextInt() - 1, z = sc.nextInt();
			adj[x].add(new Edge(y, z));
			adj[y].add(new Edge(x, z));
		}
		
		boolean[][] visited = new boolean[n][c+1];
		LinkedList<Q> q = new LinkedList<Q>();
		q.add(new Q(a, 0, 0));
		while (!q.isEmpty()) {
			Q cur = q.pop();
			
			if (cur.fuel > c)
				cur.fuel = c;
			if (cur.fuel < 0 || visited[cur.node][cur.fuel])
				continue;
			if (cur.node == b) {
				System.out.println(cur.dist);
				break;
			}
			
			visited[cur.node][cur.fuel] = true;
			q.add(new Q(cur.node, cur.fuel + p[cur.node], cur.dist + 1));
			for (Edge e : adj[cur.node])
				q.add(new Q(e.v, cur.fuel - e.w, cur.dist + 1));
		}
		
		sc.close();

	}
	
	static class Edge {
		int v, w;
		
		public Edge(int v, int w) {
			this.v = v;
			this.w = w;
		}
	}
	
	static class Q {
		int node, fuel, dist;
		
		public Q(int node, int fuel, int dist) {
			this.node = node;
			this.fuel = fuel;
			this.dist = dist;
		}
	}

}

File: archive/2019/solutions/MilkyWay.py
class Vertex:
    def __init__(self):
        self.adj = []
    def addEdge(self, v, wt):
        self.adj.append((v, wt))

n, m, c = [int(i) for i in input().split()]
a, b = [int(i)-1 for i in input().split()]
p = [int(i) for i in input().split()]

graph = [Vertex() for i in range(n)]
for i in range(m):
    x, y, z = [int(i) for i in input().split()]
    graph[x-1].addEdge(y-1, z)
    graph[y-1].addEdge(x-1, z)

visited = [[False for i in range(c+1)] for j in range(n)]
q = [(a, 0, 0)]
while q:
    node, fuel, dist = q[0]
    q.pop(0)

    if fuel > c:
        fuel = c
    if fuel < 0 or visited[node][fuel]:
        continue
    if node == b:
        print(dist)
        break

    visited[node][fuel] = True
    q.append((node, fuel + p[node], dist + 1))
    for e in graph[node].adj:
        q.append((e[0], fuel - e[1], dist + 1))
